source/common/tcp/conn_info_linux.cc

```diff
@@ -0,0 +1,25 @@
+#if !defined(__linux__)
+#error "Linux platform file is part of non-Linux build."
+#endif
+
+#include "common/tcp/conn_info.h"
+
+#include "envoy/common/platform.h"
+#include "envoy/network/socket.h"
+
+namespace Envoy {
+namespace Tcp {
+
+absl::optional<std::chrono::milliseconds>
+ConnectionInfo::lastRoundTripTime(Envoy::Network::Socket* socket) {
+  struct tcp_info ti;
+  socklen_t len = sizeof(ti);
+  if (!SOCKET_FAILURE(socket->getSocketOption(IPPROTO_TCP, TCP_INFO, &ti, &len).rc_)) {
```

Comment:
We should do platform specific handling only in the IoHandle and os_syscall_impl. Probably IPPROTO_TCP and TCP_INFO would need to be converted to Envoy specific macros to be defined [here](https://github.com/envoyproxy/envoy/blob/master/include/envoy/common/platform.h) since Windows does not support them. 

A quick search suggests that rtt can't be retrieved using getsockopt on Windows, so either map ENVOY_TCP_INFO to something that will return an error or change os_syscall_impl to convert the getsockopt call into something that could do the job.

Reply:
What were you envisioning the API would look like? A new `lastRoundTripTime` method in [`//include/envoy/api/os_sys_calls_linux.h`](https://github.com/envoyproxy/envoy/blob/master/include/envoy/api/os_sys_calls_linux.h)?

I don't feel like this fits the pattern of `os_syscall_impl` since it is built using a system call rather than being a system call itself.

I modeled this code after the pattern I saw for [`//source/server/options_impl_platform_linux.cc`](https://github.com/envoyproxy/envoy/blob/master/source/server/options_impl_platform_linux.cc).

Reply:
For linux, in particular, there's nothing more to do beyond the function you've written. But, instead of having platform specific implementations of this function, i.e., `conn_info_default.cc` and `conn_info_linux.cc`, it would be better if we pushed this complexity down into the platform specific `Socket`/`IoHandle` code. Since we don't have platform specific `IoHandle` implementations, the wrappers we keep for socket specific syscalls would be the next best place to do this, if we have to. 

In other words, if we're able to ensure that `socket->getSocketOption(IPPROTO_TCP, ENVOY_TCP_INFO, &ti, &len)` runs independent of platform (actually, `IoHandle` implementation), we simplify the rest of the code. Given that this already works on linux and macos seems to support `struct tcp_info` (haven't tried it though), the only thing left is to make sure that it works on windows. Problem is that `TCP_INFO` and `struct tcp_info` are probably not defined. However, since you don't seem to be interested in the result for windows, one option would be to `#define ENVOY_TCP_INFO  -1` or some value that's not [here](https://docs.microsoft.com/en-us/windows/win32/winsock/ipproto-tcp-socket-options) and `struct tcp_info { uint32_t empty; }` in [here](https://github.com/envoyproxy/envoy/blob/master/include/envoy/common/platform.h) when the platform is windows. For linux and macos only `#define ENVOY_TCP_INFO TCP_INFO`. That should ensure that the code builds on windows and the call returns an error which `lastRoundTripTime` will ignore. See `ENVOY_TCP_BACKLOG_SIZE` which is used in a similar way. 

Also, keep in mind that a custom `IoHandle` implementation on macos/windows could be able to return the connection's rtt if it doesn't rely on OS sockets.

Reply:
+1. I think adding the lastRoundTripTime to the `ConnectionSocket` and then plumbing it into the `IoHandle` will fit current architecture better.  This method should be implemented in the `IoSocketHandleImpl` class. It already have `#ifdef`s based on platform. You can just ifdef it based on TCP_INFO.

Reply:
Sounds good, thanks for the comments Florin and Yan. I'll pick this back up in a bit and do the refactoring suggested.

Reply:
I attempted to make the changes as you have suggested. Can you tell me if the new implementation is what you're expecting? If so, I can move the tests as well.

Reply:
LGTM. Thanks!